### PR TITLE
PageFind for users in China

### PR DESCRIPTION
**Pull request by cjyabraham**
First, this PR has PageFind search available for [all users OUTSIDE of China](https://github.com/cjyabraham/website/commit/3c0442eab47dfa2f91e1c08c102eaff4776e325e). This is just temporary to facilitate testing and we will revert this before merging the PR. The goal is to show PageFind to all users IN China, as a replacement for Bing, which has been [costing CNCF a lot of money recently](https://github.com/kubernetes/website/issues/44475).

Could someone help integrate this into the Makefile properly. I’m not sure where it should be applied and how to integrate it. You can see what I’ve done to get PageFind to run in [the netlify.toml commands](https://github.com/kubernetes/website/pull/46768/files#diff-ab8f79b68b7adff7a07db953bf453f3c5aa6ade98d2b1b67d8432b36392489ed) but that may not be the right place for them.

[Searches with PageFind](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=cloud) can be compared against [the current site](https://kubernetes.io/search/?q=cloud). If they are adequate, then we can get this deployed soon to remove Bing from the site. We should specifically test searches in Chinese but also test English and other languages.

We can enhance the PageFind search functionality in the future with [filters](https://pagefind.app/docs/filtering/), [sorting](https://pagefind.app/docs/sorts/), etc. If the PageFind results become better than those of Google Custom Search, we can have it serve all users, not just those in China.

I pulled ideas from #44655 when creating this PR but started from scratch as I couldn’t get that PR branch to run locally. I also kept the scope narrower and less ambitious, limiting PageFind search to just Chinese users while keeping Google Custom Search for everyone else. This was so that we could address the immediate need of getting off Bing as quickly as possible.

FYI this is the command I use to test this locally:
```
hugo && npx -y pagefind --site public --serve
```

### CODE REVIEW

**Comment by sftim** [comment]
Hi @cjyabraham 

If I go to https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/docs/home/ and search for kubectl 命令行工具, I get pages in English. To be fair, I'm using a browser that prefers English but I was looking at the Chinese docs. This may well surprise users.

https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=kubectl%20%E5%91%BD%E4%BB%A4%E8%A1%8C%E5%B7%A5%E5%85%B7

**Comment by cjyabraham** [comment]
Actually now that I think of it, it's [a known issue with doscy](https://github.com/google/docsy/issues/1558) which [I have already submitted a fix for in this PR](https://github.com/google/docsy/pull/1904). Since that's not merged yet, we can reapply the fix locally within this site...

**Comment by cjyabraham** [comment]
Ok, I've fixed that issue so now the site should correctly keep the user on the search page of the same language they were on. This is going to be a problem, however, for languages that don't have search pages, like French, so perhaps we should quickly add content files for them?

**Review by sftim — DISMISSED** [review]
We shouldn't add “localized” pages where the content is in English. Let's find another way to make that work.

For example, change the search form to post to the localized search page *iff* there is one; if there isn't post to the English search page.

**Comment by cjyabraham** [comment]
> We shouldn't add “localized” pages where the content is in English. Let's find another way to make that work.
> 
> For example, change the search form to post to the localized search page _iff_ there is one; if there isn't post to the English search page.

The only string that needs translating on the new search pages are the words "Search Results" which appears as the `<title>` for the page. Everything else is in the correct language. Would you be open to us generating a translation for these strings? That way we would be able to immediately provide search results in the languages in question.

**Comment by sftim** [comment]
> The only string that needs translating on the new search pages are the words "Search Results" which appears as the `<title>` for the page. Everything else is in the correct language. Would you be open to us generating a translation for these strings? That way we would be able to immediately provide search results in the languages in question.

I don't recommend it; you will need to get 14 approvals (base approval plus 13 translation signoffs), and this is not going to be straightforward.

See https://github.com/kubernetes/website/pull/46768#pullrequestreview-2110222558 for the approach I recommend instead. If you're not sure how to code that, it's OK to ask for help with this part - #sig-docs on Slack is another good place to take that conversation.

**Comment by cjyabraham** [comment]
> I don't recommend it; you will need to get 14 approvals (base approval plus 13 translation signoffs), and this is not going to be straightforward.
> 
> See [#46768 (review)](https://github.com/kubernetes/website/pull/46768#pullrequestreview-2110222558) for the approach I recommend instead. If you're not sure how to code that, it's OK to ask for help with this part - #sig-docs on Slack is another good place to take that conversation.

Ok, I've implemented your suggestion.

**Comment by sftim** [comment]
Please squash out the commits that touched localized files @cjyabraham - we can then take those teams out of the approval process.

File: layouts/partials/search-input.html
Comment (sftim) [inline layouts/partials/search-input.html:37]: This is a big fragile; a better way is to find the site-relative URL of the search page for [English | the primary localization].
But, it'll work for now.

File: layouts/docs/baseof.html
Comment (sftim) [inline layouts/docs/baseof.html:25]: Can we also explicitly pass in the page title here? The `h1` element might otherwise get ignored by PageFind.

File: static/js/search.js
Comment (sftim) [inline static/js/search.js:77]: (nit / doesn't block merge)
A tidier way to do this is only if either:
- the user clicks on a PageFind link
- the user submits the form

To see the problem, edit a search input and then try navigating back.

**Comment by cjyabraham** [comment]
> Please squash out the commits that touched localized files @cjyabraham - we can then take those teams out of the approval process.

Ok, done.

File: layouts/docs/baseof.html
Comment (cjyabraham) [reply layouts/docs/baseof.html:25]: Is there some reason why you think we need to override this automatic functionality? I think PageFind is set up to use the first `h1` element as the title for the page and it seems to be working correctly as is. If you [search for Autoscaling Workloads](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=Autoscaling%20Workloads), for example, you'll see the [Autoscaling Workloads page](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/autoscaling/) correctly listed in the search results.

File: static/js/search.js
Comment (cjyabraham) [reply static/js/search.js:77]: It is currently set to perform this on form submit. This happens automatically every time the search text is changed in the input box. The updated url could be useful in the case the user wants to share the search results.

File: layouts/partials/search-input.html
Comment (cjyabraham) [reply layouts/partials/search-input.html]: Ok, I've made it a relative url so that it will work on local instances.

**Comment by sftim** [comment]
Before we approve it, I'd also like to see this PR squashed ready for merge.

File: static/js/search.js
Comment (sftim) [reply static/js/search.js:77]: You see the snag, though, right? If you have typed a term like PersistentVolume, and you want to navigate back, you now must click 16 times. Each keystroke adds a history entry.

**Comment by cjyabraham** [comment]
> Before we approve it, I'd also like to see this PR squashed ready for merge.

So do you want me to flip the switch so that PageFind is now only accessible to users in China? Or is there still some testing that needs to happen for approvals? Also, what's the advantage of squashing before merge? Can't that happen during the merge operation?

**Comment by sftim** [comment]
Squashing for merge will turn the many commit messages into a more readable summary of the change. That really helps people who come back to look at the code in the future, and who want to understand the thinking behind the change(s).

Using `/label tide/merge-method-squash` doesn't rewrite the commit messages; Tide isn't capable of summarizing text.

You should squash after this is flipped to only serve to Chinese visitors. You could squash before, as well.

**Comment by sftim** [comment]
The next thing this needs is:
- testing by people who can evaluate the Chinese-language search (many users in China will use that localization)
- testing by people who can evaluate the English-language search (some users in China will use that localization)

**Comment by windsonsea** [comment]
The preview looks good from Shanghai, China:

- [search zh pages](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=pod)
- [search en pages](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=pod)

**Comment by cjyabraham** [comment]
> The preview looks good from Shanghai, China:
> 
> * [search zh pages](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=pod)
> * [search en pages](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=pod)

@windsonsea can you confirm that you're getting PageFind search results and not Google Search results? I currently have flipped the logic so that those _outside_ China get PageFind and those _inside_ get Google. But maybe you're on a VPN?

**Comment by sftim** [comment]
More feedback is welcome

**Comment by cjyabraham** [comment]
@sftim What does this still need before we can push it live?

**Comment by sftim** [comment]
> What does this still need before we can push it live (apart from me switching the IP check algo back to normal and squashing commits)?

It needs a review (anyone representing SIG Docs) and an approval (SIG leadership). Evidence that this works for users is also helpful.

File: static/js/search.js
Comment (cjyabraham) [reply static/js/search.js]: Ok, I've fixed this.

**Comment by cjyabraham** [comment]
> I also **recommend** squashing it to fewer commits.

Ok, done.

**Comment by sftim** [comment]
Commits fd7d531f69b31d2d340bfe5d745bafb7b33c0926 and bede881988db62e829f139c9ed4e8e6590d24915 don't look right.

I can make a branch that has the commits I think you'd want, if that's useful @cjyabraham?

**Comment by cjyabraham** [comment]
> Commits [fd7d531](https://github.com/kubernetes/website/commit/fd7d531f69b31d2d340bfe5d745bafb7b33c0926) and [bede881](https://github.com/kubernetes/website/commit/bede881988db62e829f139c9ed4e8e6590d24915) don't look right.
> 
> I can make a branch that has the commits I think you'd want, if that's useful @cjyabraham?

Yes, I'm not sure how those commits got in this branch. Sure, please go ahead and do whatever is needed. Thanks.

File: layouts/docs/baseof.html
Comment (sftim) [reply layouts/docs/baseof.html:25]: Try searching:
https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=Pod
https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=%E5%91%BD%E4%BB%A4%E8%A1%8C%E5%B7%A5%E5%85%B7%20(kubectl)
https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=%E4%BD%BF%E7%94%A8+kubeadm+%E5%BC%95%E5%AF%BC%E9%9B%86%E7%BE%A4

For each of those, there is a page with the title exactly matching the search string, and it'd be be a good result to serve. Top of the result list ideally. For [命令行工具 (kubectl)](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/docs/reference/kubectl/) the page is not even in the first page of results.

**Comment by sftim** [comment]
See https://github.com/kubernetes/website/pull/46768#discussion_r1650090961 for some feedback. I don't think it blocks a merge, but it could have an effect on end-user experience.

File: layouts/docs/baseof.html
Comment (cjyabraham) [reply layouts/docs/baseof.html:25]: I guess there are a number of factors that go into the page ranking. I've looked for a way to increase the weight on the page title in the ranking, however, [it doesn't look like pagefind supports this currently](https://pagefind.app/docs/ranking/).

Comment (cjyabraham) [reply layouts/docs/baseof.html:25]: Are there equivalent to quote marks in Chinese? When I wrap the search string in quote marks, at least in the English version, it produces results with the matching page title higher up: https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=%22Command+line+tool+%28kubectl%29%22

Comment (cjyabraham) [reply layouts/docs/baseof.html:25]: Ok, I've increased [the weight on the h1 element](https://pagefind.app/docs/weighting/#default-rankings) in pagefind results. Now the listing of the page with the search term as title are higher up the listing (although still not on the first page). When searching for "命令行工具 (kubectl)", the page with that title has gone from rank 11 to [9](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/zh-cn/search/?q=%E5%91%BD%E4%BB%A4%E8%A1%8C%E5%B7%A5%E5%85%B7%20(kubectl)).

Is there a way to put a search string in quote for the Chinese language? I notice when I do that for the English phrase "Command line tool (kubectl)", the page with that title goes from rank 6 to [rank 3](https://deploy-preview-46768--kubernetes-io-main-staging.netlify.app/search/?q=%22Command+line+tool+%28kubectl%29%22), as you'd hope.

Comment (sftim) [reply layouts/docs/baseof.html:25]: I think that's enough tweaking that we can ship this / my concern is addressed.

**Comment by sftim** [comment]
Ready for more feedback from people likely to use this (ie, the far side of China's nation-level content filtering).

**Comment by mengjiao-liu** [comment]
> See [#46768 (comment)](https://github.com/kubernetes/website/pull/46768#discussion_r1650090961) for some feedback. I don't think it blocks a merge, but it could have an effect on end-user experience.

LGTM from China.

**Comment by sftim** [comment]
/lgtm cancel
(need to swap the sense of the change, now it's testing)

But this does actually **LGTM**

@cjyabraham please rebase out or revert the commit _pretend we are in China to test this_, and then we may be able to get further testing done by users in China.

**Comment by cjyabraham** [comment]
> @cjyabraham please rebase out or revert the commit _pretend we are in China to test this_, and then we may be able to get further testing done by users in China.

It was not a clean revert of one commit as a couple changes were needed, so I made a new commit.

**Comment by cjyabraham** [comment]
> (happy for me to squash this down @cjyabraham ?)

Please do. You seem to be particularly good at this. And if you know of a good resource where I can learn more about it then please send it over :)

**Comment by cjyabraham** [comment]
@sftim what is the next step here?

**Comment by sftim** [comment]
> what is the next step here?

This PR is ready for review. It needs an LGTM, _ideally_ a bit more evidence that it doesn't break things for users not behind state level censorship, and a website approver must approve the change.

If you use Kubernetes Slack, you can nudge #sig-docs about reviews and approval.

**Comment by cjyabraham** [comment]
I'm not on Kubernetes slack. Can we tag people here that need to look at this?

**Comment by sftim** [comment]
> Can we tag people here that need to look at this?

Sure. Write a message and tag `@kubernetes/sig-docs-leads` in your message.

**Comment by cjyabraham** [comment]
Can we have this PR reviewed please @kubernetes/sig-docs-leads

**Comment by nate-double-u** [comment]
I've pinged folks in the Kubernetes Slack channels #sig-docs, #sig-docs-localizations, and #kubernetes-docs-zh to see if they can help review.

**Comment by sftim** [comment]
/hold

This is also serving PageFind to users outside of China

**Comment by cjyabraham** [comment]
@sftim it seems that your rebase operation [removed my last commit](https://github.com/kubernetes/website/compare/02663cdee83deb30f8fa5629a95e504006b06d3a..c86a1576dedaa56d1fce1ef0cbb7dda149477ba2). Could you undo that?

**Comment by sftim** [comment]
I'll update this PR when I get a moment.

**Comment by sftim** [comment]
The PageFind functionality is only visible to users in China. However, you can check this out locally and try it by editing the `search.js` file; see https://github.com/kubernetes/website/compare/c86a1576dedaa56d1fce1ef0cbb7dda149477ba2..a4191d94ea52f8c4935c144f0d89d779b16f5812

**Comment by natalisucks** [comment]
/hold

@kubernetes/steering-committee howdy steering! this change has come about out of the CNCF's need to replace Bing for users in China, and our zh-localization community, alongside other folks in SIG Docs, have reviewed this change to ensure it's doing what it describes. we want to check if this needs to be run by you before merging – thanks!

**Comment by mrbobbytables** [comment]
LGTM on my side, I don't think it needs steering sign off :+1:

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/website/pull/46768#issuecomment-2207446787 Agree.

**Comment by BenTheElder** [comment]
Feel free to unhold, or wait for a simple majority of steering (routine business rules would apply if a vote were needed for something like this), but I pretty strongly think this does not need to wait for steering and is delegated to SIG Docs. 

(re: delegation, maybe reach out to SIG K8s Infra if we have new paid SaaS accounts to manage but AFAICT this is not the case here and still wouldn't involve steering)

**Comment by natalisucks** [comment]
sick, thanks @mrbobbytables and @BenTheElder!
/hold cancel
/lgtm

**Comment by nate-double-u** [comment]
Thanks for this @cjyabraham & @sftim, and thanks to all the reviewers!

/approve

**Comment by SayakMukhopadhyay** [comment]
Hey @cjyabraham , I am working on aligning the search closer to Docsy as per #41171 and I wanted to understand the reasoning behind deleting the search from the sidebar for pagefind.

I do notice that the default search bar in the search results page is replace by the pagefind searchbar. Is this piece of code meant for removing the search bar in the search results?

**Comment by SayakMukhopadhyay** [comment]
I was wondering why the `.td-sidebar__search` class was used to find the element. Cause its now having some side effects in #49724. The search in the sidebar of docs/blogs page don't show up in that PR for pagefind users. This is because the search in the sidebar uses the same class.

I did some check and it seems a [change](https://github.com/kubernetes/website/pull/49724/files#diff-f1ed5217add3987eed77d3932f72f3096c4780c9353905cd8325f2495e0a14b1R7-R9) I did in that PR caused the search bar in the search results no longer have the `.td-sidebar__search` class. Anyway, thanks for jumping in @sftim , I will push a commit in that PR to fix it for pagefind users.

**Pull request merged**
